lib/bulk_insert/worker.rb

```diff
@@ -76,14 +77,21 @@ def after_save(&block)
     def save!
       if pending?
         @before_save_callback.(@set) if @before_save_callback
-        compose_insert_query.tap { |query| @connection.execute(query) if query }
+        execute_query
         @after_save_callback.() if @after_save_callback
         @set.clear
       end
 
       self
     end
 
+    def execute_query
+      if query = compose_insert_query
+        result_set = @connection.execute(query)
+        @result_sets.push(result_set)
```

Comment:
There can be multiple result sets because bulk_insert inserts records in batches of 500. For example, if you have to insert 1000 records, `save!` would be called for the first 500 and the result set containing the ids would be pushed on to `@result_sets`. Same thing for the second 500.

The reason for using an array of result sets instead of an array of the results is so that the worker doesn't have to build a new array of 1000 items; it can just build an array of 2 result sets returning from the query. Of course, if we care about the returned ids, we'll still have to use `@result_sets` go through all 1000 results. But most of the time we don't care about the returned ids.

Reply:
this is a great solution. I love that it doesn't break the existing interface and therefore reduces the risk of this change we are making. 

My last question is, if the inserts happen in series of 500, how can the consumer of the worker know that there are no outstanding series to run. I.e. if you call bulk insert and immediately call worker.result_sets, how we be sure bulk insert has finished. 

Maybe the answer is that the worker is not actually executing in an async manner, and that when bulk_insert returns, all queries have already been executed

Reply:
ahh yes, i just dug into the implementation of worker. It appears as then when save! is called it really does execute and then return. batches are run on the fly as the number of records in the worker increases.

This works great then!